DigitalLearningSolutions.Data/Helpers/DateHelper.cs

```diff
@@ -3,10 +3,81 @@
     using System;
     using System.Collections.Generic;
     using System.Linq;
+    using DigitalLearningSolutions.Data.Enums;
+    using DigitalLearningSolutions.Data.Models.TrackingSystem;
 
     public static class DateHelper
     {
-        public static IEnumerable<(int Month, int Year)> GetMonthsAndYearsBetweenDates(DateTime startDate, DateTime endDate)
+        public static IEnumerable<DateInformation> GetPeriodsBetweenDates(
+            DateTime startDate,
+            DateTime endDate,
+            ReportInterval interval
+        )
+        {
+            switch (interval)
+            {
+                case ReportInterval.Days:
+                    return GetDaysBetweenDates(startDate, endDate);
+                case ReportInterval.Weeks:
+                    return GetWeeksBetweenDates(startDate, endDate);
+                case ReportInterval.Months:
+                    return GetMonthsBetweenDates(startDate, endDate);
+                case ReportInterval.Quarters:
+                    return GetQuartersBetweenDates(startDate, endDate);
+                default:
+                    return GetYearsBetweenDates(startDate, endDate);
+            }
+        }
+
+        private static IEnumerable<DateInformation> GetDaysBetweenDates(
+            DateTime startDate,
+            DateTime endDate
+        )
+        {
+            var diffInDays = (endDate - startDate).Days;
+            var daysDifferenceEnumerable = Enumerable.Range(0, diffInDays + 1);
+
+            return daysDifferenceEnumerable.Select(
+                d =>
+                {
+                    var day = startDate.AddDays(d);
+                    return new DateInformation
+                    {
+                        Interval = ReportInterval.Days,
+                        Date = new DateTime(day.Year, day.Month, day.Day)
+                    };
+                }
+            );
+        }
+
+        private static IEnumerable<DateInformation> GetWeeksBetweenDates(
+            DateTime startDate,
+            DateTime endDate
+        )
+        {
+            var referenceDate = new DateTime(1905, 1, 1);
+
+            var diffInWeeks = (endDate - referenceDate).Days / 7;

```

Comment:
I'm replicating the function of the original SQL procedure, which sorts things into a consistent set of weekly slots starting on Jan 1st 1905 (a Sunday), rather than shifting the week boundary based on the dates chosen.